pkg/storegateway/indexheader/index/postings.go

```diff
@@ -0,0 +1,514 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/tsdb/index/postings.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors.
+
+package index
+
+import (
+	"fmt"
+	"hash/crc32"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/index"
+
+	stream_encoding "github.com/grafana/mimir/pkg/storegateway/indexheader/encoding"
+)
+
+const postingLengthFieldSize = 4
+
+type PostingOffsetTable interface {
+	// PostingsOffset returns a list of offsets for labels named name with value in values.
+	// TODO: can this be simplified to only handle a single value?
+	// StreamBinaryReader.PostingsOffset only ever calls this method with a single value, although there's a comment
+	// in there about taking advantage of retrieving multiple values at once.
+	PostingsOffset(name string, values ...string) ([]index.Range, error)
+
+	// LabelValues returns a list of values for the label named name that match filter.
+	LabelValues(name string, filter func(string) bool) ([]string, error)
+
+	// LabelNames returns a list of all label names in this table.
+	LabelNames() ([]string, error)
+
+	// ForEachLabelName calls f for each label name in this table.
+	// If f returns a non-nil error for any label name, ForEachLabelName immediately returns that error
+	// and does not call f again.
+	ForEachLabelName(f func(name string) error) error
+
+	// LabelNameCount returns a value which is at least as large as the number of
+	// label names present in this table.
+	LabelNameCount() int
+}
+
+type PostingOffsetTableV1 struct {
+	// For the v1 format, labelname -> labelvalue -> offset.
+	postings map[string]map[string]index.Range
+}
+
+func NewPostingOffsetTable(factory *stream_encoding.DecbufFactory, tableOffset int, indexVersion int, indexLastPostingEnd uint64, postingOffsetsInMemSampling int) (PostingOffsetTable, error) {
+	if indexVersion == index.FormatV1 {
+		return newV1PostingOffsetTable(factory, tableOffset, indexLastPostingEnd)
+	} else if indexVersion == index.FormatV2 {
+		return newV2PostingOffsetTable(factory, tableOffset, indexLastPostingEnd, postingOffsetsInMemSampling)
+	}
+
+	return nil, fmt.Errorf("unknown index version %v", indexVersion)
+}
+
+func newV1PostingOffsetTable(factory *stream_encoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64) (PostingOffsetTable, error) {
```

Comment:
Nit, should return a concrete type:

```suggestion
func newV1PostingOffsetTable(factory *stream_encoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64) (*PostingOffsetTableV1, error) {
```